### PR TITLE
Fix pickling of KFoldLoop

### LINKED ISSUE #12439: KFold example with `ddp_spawn` crash
## 🐛 Bug

When using the cross validation loop from the example `pl_examples/loop_examples/kfold.py` with `ddp_spawn` strategy, it encounter a SIGABRT Exception and the program crash.

### To Reproduce
```Python
import os
from pytorch_lightning import seed_everything, Trainer

from pl_examples.loop_examples.kfold import KFoldLoop, LitImageClassifier, MNISTKFoldDataModule


def run():
    seed_everything(42)
    model = LitImageClassifier()
    datamodule = MNISTKFoldDataModule()

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        strategy="ddp_spawn",
    )
    internal_fit_loop = trainer.fit_loop
    trainer.fit_loop = KFoldLoop(5, export_path="./")
    trainer.fit_loop.connect(internal_fit_loop)
    trainer.fit(model, datamodule=datamodule)


if __name__ == "__main__":
    run()
```

### Expected behavior

Training without exception

### Environment
* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.20.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0
	- pytorch-lightning: 1.6.0rc0
	- tqdm:              4.60.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         
	- python:            3.8.9
	- version:           #1 SMP Fri Jan 14 13:59:45 UTC 2022

### Additional context

I am working on a fix.

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by niberger**
## What does this PR do?

Make the `KFoldLoop` example class serializable via pickling.
It allow the use of this example with the `ddp_spawn` strategy.
Without this fix, the deserialization of `KFoldLoop` crash by entering an endless recursive call of `__getattr__("fitloop")`.

No test written as this only improve an example.

Fixes #12439

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/loop_examples/kfold.py [hunk 1]
```diff
@@ -239,6 +239,9 @@ def __getattr__(self, key) -> Any:
             return getattr(self.fit_loop, key)
         return self.__dict__[key]
 
+    def __setstate__(self, state):
+        self.__dict__.update(state)
```
Comment (awaelchli) [inline pl_examples/loop_examples/kfold.py:243]: is this because we have overridden `__getattr__`?

File: pl_examples/loop_examples/kfold.py [hunk 2]
```diff
@@ -239,6 +239,9 @@ def __getattr__(self, key) -> Any:
             return getattr(self.fit_loop, key)
         return self.__dict__[key]
 
+    def __setstate__(self, state):
```
Comment (awaelchli) [inline pl_examples/loop_examples/kfold.py:242]:
```suggestion
    def __setstate__(self, state: Dict[str, Any]) -> None:
```

**Review by awaelchli — APPROVED** [review]

**Review by krshrimali — APPROVED** [review]
Thanks, @niberger for the contribution! 🚀 LGTM!

**Review by akihironitta — APPROVED** [review]
LGTM!

**Review by carmocca — APPROVED** [review]
